### ISSUE TITLE
Puma caught this error: undefined method `reject' for nil:NilClass (NoMethodError)

### ISSUE THREAD

**Issue by wbotelhos** [rack/rack#1576]
Updating rack from `2.1.2` to `2.2.0` breaks `Rack::Attack`.

```
Rack::Attack.blocklist do |req|
  Rack::Attack::Fail2Ban.filter("pentesters-#{req.ip}", maxretry: 1, findtime: 10.minutes, bantime: 5.seconds) do # 82
    matched = req.path =~ regex
  end
end
```

Error:

```
Puma caught this error: undefined method `reject' for nil:NilClass (NoMethodError)
/Users/wbotelhos/.rvm/gems/ruby-2.6.5@danca/gems/rack-2.2.0/lib/rack/request.rb:623:in `reject_trusted_ip_addresses'
/Users/wbotelhos/.rvm/gems/ruby-2.6.5@danca/gems/rack-2.2.0/lib/rack/request.rb:363:in `ip'
/Users/wbotelhos/workspace/danca/config/initializers/rack_attack.rb:82:in `block in <class:Attack>'
/Users/wbotelhos/.rvm/gems/ruby-2.6.5@danca/gems/rack-attack-6.2.2/lib/rack/attack/check.rb:14:in `matched_by?'
/Users/wbotelhos/.rvm/gems/ruby-2.6.5@danca/gems/rack-attack-6.2.2/lib/rack/attack.rb:95:in `block in blocklisted?'
/Users/wbotelhos/.rvm/gems/ruby-2.6.5@danca/gems/rack-attack-6.2.2/lib/rack/attack.rb:95:in `any?'
/Users/wbotelhos/.rvm/gems/ruby-2.6.5@danca/gems/rack-attack-6.2.2/lib/rack/attack.rb:95:in `blocklisted?'
/Users/wbotelhos/.rvm/rubies/ruby-2.6.5/lib/ruby/2.6.0/forwardable.rb:230:in `blocklisted?'
/Users/wbotelhos/.rvm/gems/ruby-2.6.5@danca/gems/rack-attack-6.2.2/lib/rack/attack.rb:164:in `call'
/Users/wbotelhos/.rvm/gems/ruby-2.6.5@danca/gems/puma-4.3.1/lib/puma/configuration.rb:228:in `call'
/Users/wbotelhos/.rvm/gems/ruby-2.6.5@danca/gems/puma-4.3.1/lib/puma/server.rb:681:in `handle_request'
/Users/wbotelhos/.rvm/gems/ruby-2.6.5@danca/gems/puma-4.3.1/lib/puma/server.rb:472:in `process_client'
/Users/wbotelhos/.rvm/gems/ruby-2.6.5@danca/gems/puma-4.3.1/lib/puma/server.rb:328:in `block in run'
/Users/wbotelhos/.rvm/gems/ruby-2.6.5@danca/gems/puma-4.3.1/lib/puma/thread_pool.rb:134:in `block in spawn_thread'
```

**Comment by ioquatix** [comment]
I will take a look.

**Comment by jeremyevans** [comment]
Looks like we should change the `else` to `elsif forwarded_ips = self.forwarded_for` in `Request#ip`.

**Comment by ioquatix** [comment]
Yeah, I'm just looking at that. The problem is I see we have two different interpretations of `REMOTE_ADDR`...

```
split_header(get_header('REMOTE_ADDR'))
```

which implies it might be a comma separated list of addresses and a few lines later:

```
|| get_header("REMOTE_ADDR")
```

which just returns the entire header?

**Comment by jeremyevans** [comment]
I'm guessing this was caused by 290523f67cc43c5847b2be2d12964d1232061fe1. `forwarded_ips` was previously always an array, and it changed to possibly being `nil` in that commit.

**Comment by ioquatix** [comment]
Yes, that's true, but why on line 1 do we treat REMOTE_ADDR as a list of addresses and the at the end just return it verbatim?

**Comment by jeremyevans** [comment]
I guess because we want to return a trusted ip address instead of nothing?  If `REMOTE_ADDR` consistents only of trusted ip addresses, it could result in `remote_addrs` being empty.

**Comment by ioquatix** [comment]
Yes, but on the first line we are considering `REMOTE_ADDR` as a comma separated list of addresses, but on the last line we are just returning it verbatim.

**Comment by ioquatix** [comment]
Let me make some failing specs to start with. Then we can figure out a solution.

**Comment by ioquatix** [comment]
Even without the bug I introduced, this method has issues:

```
    res = mock.get '/', 'REMOTE_ADDR' => '127.0.0.1,127.0.0.1'
```

returns `127.0.0.1,127.0.0.1`.

**Comment by ioquatix** [comment]
@wbotelhos can you try running against rack master and see if the issue is fixed?

**Comment by kibenimatik** [comment]
@ioquatix in my case issue has gone in rack master

**Comment by ioquatix** [comment]
Okay great, I guess we will release 2.2.1 soon. Perhaps we will wait a day to see if any other issues crop up.

**Comment by ioquatix** [comment]
Okay, released in v2.2.1 just now.

**Comment by wbotelhos** [comment]
Everything working perfectly, @ioquatix . Thank you! <3
